tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,267 @@
+        let mut cache = self.cache.lock().await;
+
+        for (address, channel) in self.channels() {
+            let mut client = NodeCtlSvcClient::new(channel.clone())
+                .accept_compressed(CompressionEncoding::Gzip)
+                .send_compressed(CompressionEncoding::Gzip);
+
+            let response = match client.get_ident(()).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            // node is reachable and answered
+            answer = true;
+            if response.status != NodeStatus::Alive as i32 {
+                // node did not join the cluster yet.
+                continue;
+            }
+
+            // cache the channel for later usage
+            cache.insert(
+                response
+                    .node_id
+                    .map(|id| PlainNodeId::from(id.id))
+                    .expect("id must be set"),
+                channel,
+            );
+
+            let req = GetMetadataRequest {
+                kind: MetadataKind::NodesConfiguration.into(),
+                sync: false,
+            };
+
+            let mut response = match client.get_metadata(req).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            let nodes_configuration =
+                StorageCodec::decode::<NodesConfiguration, _>(&mut response.encoded)
+                    .map_err(|err| ConnectionInfoError::InvalidNodesConfiguration(address, err))?;
+
+            if nodes_configuration.version()
+                > latest_configuration
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                latest_configuration = Some(nodes_configuration);
+            }
+        }
+
+        if !answer {
+            // all nodes has errored
+            return Err(ConnectionInfoError::NodesErrors(errors));
+        }
+
+        *config_guard = latest_configuration.clone();
+        latest_configuration.ok_or(ConnectionInfoError::MissingNodesConfiguration)
+    }
+
+    pub async fn try_each<F, T, Fut>(
+        &self,
+        role: Option<Role>,
+        mut closure: F,
+    ) -> Result<Response<T>, ConnectionInfoError>
+    where
+        F: FnMut(Channel) -> Fut,
+        Fut: Future<Output = Result<Response<T>, Status>>,
+    {
+        let nodes_config = self.get_nodes_configuration().await?;
+        let mut channels = self.cache.lock().await;
+
+        let iterator = match role {
+            Some(role) => Either::Left(nodes_config.iter_role(role)),
+            None => Either::Right(nodes_config.iter()),
+        }
+        .sorted_by(|left, right| {
```

Comment:
Nit: The left/right naming threw me for a second because the stream is of type `Either` - I'd maybe just use `a`/`b` here instead :-)